conjure-codegen/src/servers.rs

```diff
@@ -235,442 +222,482 @@ enum ReturnType<'a> {
+    let endpoint_impl = generate_endpoint_impl(ctx, def, endpoint, Style::Sync);
+    let async_endpoint_impl = generate_endpoint_impl(ctx, def, endpoint, Style::Async);
 
-    let (params, lt) = match style {
-        Style::Async => (quote!(<'a>), quote!('a)),
-        Style::Sync => (quote!(), quote!()),
-    };
-    let service = quote!(service_);
-    let mut path_params = quote!(path_params_);
-    let mut query_params = quote!(query_params_);
-    let mut headers = quote!(headers_);
-    let body = quote!(body_);
-    let response_visitor = quote!(response_visitor_);
-    let auth = quote!(auth_);
-    let response = quote!(response);
-    let mut return_type = quote!(#result<R::Output, conjure_http::private::Error>);
-    if let Style::Async = style {
-        let box_ = ctx.box_ident(def.service_name());
-        let send = ctx.send_ident(def.service_name());
-        return_type = quote! {
-            conjure_http::private::Pin<
-                #box_<dyn conjure_http::private::Future<Output = #return_type> + #send + #lt>,
-            >
-        };
+    quote! {
+        struct #name<T>(conjure_http::private::Arc<T>);
+
+        #endpoint_metadata
+        #endpoint_impl
+        #async_endpoint_impl
     }
+}
 
-    let handle_where = match style {
-        Style::Async => {
+fn generate_endpoint_metadata(
+    ctx: &Context,
+    service: &ServiceDefinition,
+    endpoint: &EndpointDefinition,
+) -> TokenStream {
+    let endpoint_name = endpoint_name(ctx, endpoint);
+
+    let some = ctx.some_ident(service.service_name());
+    let none = ctx.none_ident(service.service_name());
+    let option = ctx.option_ident(service.service_name());
+
+    let method = endpoint
+        .http_method()
+        .as_str()
+        .parse::<TokenStream>()
+        .unwrap();
+
+    let path = http_paths::parse(endpoint.http_path()).map(|segment| match segment {
+        PathSegment::Literal(lit) => quote! {
+            conjure_http::server::PathSegment::Literal(
+                conjure_http::private::Cow::Borrowed(#lit),
+            )
+        },
+        PathSegment::Parameter { name, regex } => {
+            let regex = match regex {
+                Some(regex) => {
+                    quote!(#some(conjure_http::private::Cow::Borrowed(#regex)))
+                }
+                None => quote!(#none),
+            };
             quote! {
-                where
-                    T: #lt,
-                    B: #lt,
-                    R: #lt,
+                conjure_http::server::PathSegment::Parameter {
+                    name: conjure_http::private::Cow::Borrowed(#name),
+                    regex: #regex,
+                }
             }
         }
-        Style::Sync => quote!(),
-    };
+    });
 
-    let extract_path_params = extract_path_params(ctx, endpoint, &path_params);
-    if extract_path_params.is_empty() {
-        path_params = quote!(_);
-    }
+    let template = &***endpoint.http_path();
```

Comment:
Why do we need to dereference this 3 times? The first dereference (`HttpPath::deref`) gives us a `String`. The second dereference (`String::deref`) gives us a `str`. It looks like we want `template` to be a `&str`, so I expected this to be `&**endpoint.http_path()`.

Reply:
The extra `*` is to go from `&HttpPath` to `HttpPath` I believe.

Reply:
Ah that would make sense. Thanks.